motionplan/rrtStarConnect.go

```diff
@@ -147,17 +157,56 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
 		default:
 		}
 
-		// try to connect the target to map 1
-		utils.PanicCapturingGo(func() {
-			mp.extend(rrt.maps.startMap, target, m1chan)
-		})
-		utils.PanicCapturingGo(func() {
-			mp.extend(rrt.maps.goalMap, target, m2chan)
-		})
-		map1reached := <-m1chan
-		map2reached := <-m2chan
-
-		if map1reached != nil && map2reached != nil {
+		tryExtend := func(target []referenceframe.Input) (node, node, error) {
+			// attempt to extend maps 1 and 2 towards the target
+			utils.PanicCapturingGo(func() {
+				m1chan <- nm1.nearestNeighbor(nmContext, mp.planOpts, newConfigurationNode(target), map1)
+			})
+			utils.PanicCapturingGo(func() {
+				m2chan <- nm2.nearestNeighbor(nmContext, mp.planOpts, newConfigurationNode(target), map2)
+			})
+			nearest1 := <-m1chan
+			nearest2 := <-m2chan
+			// If ctx is done, nearest neighbors will be invalid and we want to return immediately
+			select {
+			case <-ctx.Done():
+				return nil, nil, ctx.Err()
+			default:
+			}
+
+			utils.PanicCapturingGo(func() {
+				mp.constrainedExtend(ctx, map1, nearest1, newConfigurationNode(target), m1chan)
+			})
+			utils.PanicCapturingGo(func() {
+				mp.constrainedExtend(ctx, map2, nearest2, newConfigurationNode(target), m2chan)
+			})
+			map1reached := <-m1chan
+			map2reached := <-m2chan
+
+			return map1reached, map2reached, nil
+		}
+
+		map1reached, map2reached, err := tryExtend(target)
+		if err != nil {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+			return
+		}
+
+		reachedDelta := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: map1reached.Q(), EndConfiguration: map2reached.Q()})
+
+		// Second iteration; extend maps 1 and 2 towards the halfway point between where they reached
+		if reachedDelta > mp.algOpts.JointSolveDist {
+			target = referenceframe.InterpolateInputs(map1reached.Q(), map2reached.Q(), 0.5)
+			map1reached, map2reached, err = tryExtend(target)
+			if err != nil {
+				rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+				return
+			}
+			reachedDelta = mp.planOpts.DistanceFunc(&Segment{StartConfiguration: map1reached.Q(), EndConfiguration: map2reached.Q()})
+		}
+
+		// Solved
+		if reachedDelta <= mp.algOpts.JointSolveDist {
```

Comment:
this can become the else condition

Reply:
I think it needs to remain like this for the case that the previous if statement results in the maps extending enough such that reachedDelta is below the threshold. In that case we would want to run this block.